### ISSUE TITLE
Template string interpolation

### ISSUE THREAD

**Issue by chancancode** [emberjs/rfcs#623]
Just some quick notes for the idea to open up the discussion.

It's not very common, but occasionally you need to do string interpolations in Handlebars. Today, the solution is to use the `concat` helper.

```hbs
<Notification @message={{concat @user.firstName " " @user.lastName " has added you as a friend!"}} />
```

It's not so bad, but it does add some noise. It would be nicer, in my opinion, to have a string interpolation syntax.

Ideally, it would be nice if we could just use the normal string syntax:

```hbs
<Notification @message="{{@user.firstName}} {{@user.lastName}} has added you as a friend!" />
```

This would work if we are designing the language from scratch. You can always escape the mustaches if you meant for them to be literal mustaches, same as any other positions. Unfortunately, this will be a breaking change. While rare, it is possible that there are existing string literals that uses mustaches for this purpose.

The good thing about this is they are by definition static. So one option is to detect unescaped existing usages, deprecate them to clear the space, then add the interpolation feature in the next major version (of Handlebars and Ember). In the meantime, we can create an addon for apps to opt-in.

Another option is to do what JavaScript (which had the same problem and didn't have the luxury of bumping major versions) and use a different syntax for interpolated strings, say backticks:

```hbs
<Notification @message=`{{@user.firstName}} {{@user.lastName}} has added you as a friend!` />
```

**Comment by ro0gr** [comment]
is it also supposed to work within inline helpers(`{{if`), like:

```hbs
<Notification class="{{scope}} {{if this.isSomething "{{scope}}--isSomething"}}">
```
?

**Comment by chancancode** [comment]
@ro0gr Yeah, basically anything that works in "angle bracket string interpolation" today would work, I think. [`"{{"You {{"can {{"even {{"nest"}} them "}} arbitrarily"}} deep"}}"`](https://twitter.com/chancancode/status/1257933429217222656). Like maybe don't do that, but it shouldn't not work. It doesn't parse today in angle bracket position though, which seems like a bug to me.

**Comment by chancancode** [comment]
Hacked together a [very quick proof-of-concept](https://github.com/tildeio/ember-template-string-interpolation). A lot of cases doesn't work yet (e.g. if you nest the same kind of quotes it blows up), so please try it out and open issues for things that are failing!

**Comment by gossi** [comment]
I was having this case:

```
const template = hbs`...`;
const secondTemplate = hbs`the first ${template} wrapped`;
```

would that be possible with that?

**Comment by chancancode** [comment]
@gossi Its not related no. This is talking about string literals inside the template. You case should be solvable with components?

**Comment by lifeart** [comment]
@gossi https://github.com/lifeart/ember-cli-inline-template-only-components

**Comment by wagenet** [comment]
This is a good idea, but it doesn’t seem like anyone actually wants to write the RFC. Is there anything we can do to help facilitate that?

**Comment by knownasilya** [comment]
This comes up more in component design with things like a button component that has a renderless option, and you can't easily use yield as you'd use the html syntax

**Comment by chancancode** [comment]
FYI, I archived the repo for our experimental addon. We still would very much like to see the feature, but we don't have time to push it forward, and it will require some more rigorous syntax/grammar work, and on the implementation side may be non-trivial too, if we ended up needing to stop using the stock handlebars parse (which is ultimately a good/necessary thing and overdue anyway). The good news is, with modern Ember – native classes, `.gjs`/`.gts`, etc, it usually doesn't feel _too terrible_ to move the string concatenation into the JS side, whether as a getter, or a in-scope plain function, etc.

Still, for lots of reasons it would be nice to have a good first class syntax to do this in the template side directly.
